packages/interchain-token-service-api/src/abi.rs

```diff
@@ -0,0 +1,469 @@
+use alloy_primitives::{FixedBytes, U256};
+use alloy_sol_types::{sol, SolValue};
+use cosmwasm_std::{HexBinary, Uint256};
+use error_stack::{Report, ResultExt};
+use router_api::Address;
+
+use crate::{
+    error::Error,
+    primitives::{ITSMessage, ITSRoutedMessage},
+    TokenId, TokenManagerType,
+};
+
+// ITS Message payload types
+// Reference: https://github.com/axelarnetwork/interchain-token-service/blob/v1.2.4/DESIGN.md#interchain-communication-spec
+// `abi_encode_params` is used to encode the struct fields as ABI params as required by the spec.
+// E.g. `DeployTokenManager::abi_encode_params` encodes as `abi.encode([uint256, bytes32, uint256, bytes], [...])`.
+sol! {
+    enum MessageType {
+        InterchainTransfer,
+        DeployInterchainToken,
+        DeployTokenManager,
+        RoutedCall,
+    }
+
+    struct InterchainTransfer {
+        uint256 messageType;
+        bytes32 tokenId;
+        bytes sourceAddress;
+        bytes destinationAddress;
+        uint256 amount;
+        bytes data;
+    }
+
+    struct DeployInterchainToken {
+        uint256 messageType;
+        bytes32 tokenId;
+        string name;
+        string symbol;
+        uint8 decimals;
+        bytes minter;
+    }
+
+    struct DeployTokenManager {
+        uint256 messageType;
+        bytes32 tokenId;
+        uint256 tokenManagerType;
+        bytes params;
+    }
+
+    struct RoutedCall {
+        uint256 messageType;
+        /// Remote chain name.
+        /// ITS edge source contract -> ITS Hub GMP call: Set to the true destination chain name.
+        /// ITS Hub -> ITS edge destination contract: Set to the true source chain name.
+        string remote_chain;
```

Comment:
Is there a way we can not overload this field like this, and use some sort of enum? I find this confusing.

Reply:
Yeah, we can have two message types on the wire to distinguish it. The downside is that now you can have message types that are only valid in a certain direction (send to hub / receive from hub). Another option is to keep it the same on the wire but use send/receive type Rust types. But this makes decoding context sensitive which seems confusing. Discussed with SC team, and we decided to go with two separate message types